### PR TITLE
Qt 6 support

### LINKED ISSUE #648: Truncated text in the settings bar.
### Operating System

macOS

### Version of PolyMC

PolyMC Version 1.3.0

### Description of bug

When running PolyMC from my laptop with the display on my larger monitor, the text and icons in the Settings bar are weird. The icons grow outside their bounding and the text is truncated in the middle.

image

image

### Steps to reproduce

- Open PolyMC on my laptop.
- Drag it over to my second display.

### Suspected cause

My secondary screen is larger, but is has a lower pixel density and overall worse resolution, I think somehow the icons aren't scaling properly and causing the text to not fit.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Scrumplex**
~~Depends on #127~~

Closes #648

TODO:
- [x] Verify that all signals and slots work (many were renamed in Qt 6)
- [x] Make Windows x86_64 builds work (Can't find Qt platform plugin??)
- [x] Build AppImage using Qt 6

### CODE REVIEW

**Comment by DioEgizio** [comment]
this fixes #331, but 64 bit builds are broken atm

**Comment by ryanccn** [comment]
Now all checks pass

Qt 6 is generally a good idea because it supports cross-compilation and enables macOS Universal

And since now Mojang officially supports M1 for newer versions, it will be usable from the start

**Comment by Scrumplex** [comment]
About iconfix: It was first added in this commit: https://github.com/PolyMC/PolyMC/commit/ceec70e014db629aa8ba3472839cc42781d73d4c

Seems to me like this has been fixed for some time: https://bugs.kde.org/show_bug.cgi?id=344469

**Comment by Scrumplex** [comment]
IMO this is ready for review. Good luck everyone!

**Comment by Scrumplex** [comment]
Qt 6 would drop support for anything older than Windows 10. So maybe we could build x86 builds with Qt 5, while keeping our x86-64 builds on Qt 6 for Windows?

**Comment by Scrumplex** [comment]
Oh, I just remember that Windows x86-64 builds were broken last time I tried. Gonna see if it was magically fixed.

**Comment by DioEgizio** [comment]
> Qt 6 would drop support for anything older than Windows 10. So maybe we could build x86 builds with Qt 5, while keeping our x86-64 builds on Qt 6 for Windows?

Yes

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -206,7 +215,7 @@ jobs:
         shell: msys2 {0}
         run: |
           cd ${{ env.INSTALL_DIR }}
-          makensis -NOCD "${{ github.workspace }}/${{ env.BUILD_DIR }}/program_info/win_install.nsi"
+          makensis -NOCD "-DVERSION=${{ env.VERSION }}" "-DMUI_ICON=${{ github.workspace }}/program_info/polymc.ico" "-XOutFile ${{ github.workspace }}/PolyMC-Setup.exe" "${{ github.workspace }}/program_info/win_install.nsi"
```
Comment (DioEgizio) [inline .github/workflows/build.yml]: rebase oopsie?

Comment (Scrumplex) [reply .github/workflows/build.yml]: yes

Comment (DioEgizio) [reply .github/workflows/build.yml]: GH email notifications are so fucked up that this notification arrived me just now.

File: launcher/net/PasteUpload.cpp [hunk 2]
```diff
@@ -131,10 +131,20 @@ void PasteUpload::executeTask()
...
+    // This function call would be a lot shorter if we were using the latest Qt
+    connect(rep,
+            static_cast<void (QNetworkReply::*)(QNetworkReply::NetworkError)>(&QNetworkReply::errorOccurred),
+            this, &PasteUpload::downloadError);
+#else
     // This function call would be a lot shorter if we were using the latest Qt
     connect(rep,
             static_cast<void (QNetworkReply::*)(QNetworkReply::NetworkError)>(&QNetworkReply::error),
             this, &PasteUpload::downloadError);
+    connect(rep, SIGNAL(error(QNetworkReply::NetworkError)), this, SLOT(downloadError(QNetworkReply::NetworkError)));
```
Comment (flowln) [inline launcher/net/PasteUpload.cpp]: isn't this effectively the same connection as the above? o.O

File: launcher/ui/MainWindow.cpp [hunk 3]
```diff
@@ -233,7 +233,6 @@ class MainWindow::Ui
     TranslatedAction actionMods;
     TranslatedAction actionViewSelectedInstFolder;
     TranslatedAction actionViewSelectedMCFolder;
-    TranslatedAction actionViewSelectedModsFolder;
```
Comment (flowln) [inline launcher/ui/MainWindow.cpp:238]: hmmmmm isn't this basically #654? :thinking:

File: launcher/ui/dialogs/SkinUploadDialog.cpp [hunk 4]
```diff
@@ -22,10 +57,10 @@ void SkinUploadDialog::on_buttonBox_accepted()
 {
     QString fileName;
     QString input = ui->skinPathTextBox->text();
-    QRegExp urlPrefixMatcher("^([a-z]+)://.+$");
+    QRegularExpression urlPrefixMatcher(QRegularExpression::anchoredPattern("^([a-z]+)://.+$"));
     bool isLocalFile = false;
     // it has an URL prefix -> it is an URL
-    if(urlPrefixMatcher.exactMatch(input))
+    if(urlPrefixMatcher.match(input).hasMatch())  // TODO: does this work?
```
Comment (flowln) [inline launcher/ui/dialogs/SkinUploadDialog.cpp]: Seems to. Putting `aaaaaaaaa` leads to the `Skin file does not exist!` dialog outside this `if`, while putting `http://aaaaaa.com` goes to the `Using remote URLs for setting skins is not implemented yet.` dialog inside this `if`.

Yes, nice debugging tech i know :)

**Review by flowln — COMMENTED** [review]
those (c) diffs :eye: . :eye: 

jokes aside, this seems to still work fine with Qt5
the gentoo package will still depend on it for the near future, so :pensive: 

And you may want to change the PR title now :D

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -206,7 +217,7 @@ jobs:
         shell: msys2 {0}
         run: |
           cd ${{ env.INSTALL_DIR }}
-          makensis -NOCD "${{ github.workspace }}/${{ env.BUILD_DIR }}/program_info/win_install.nsi"
+           makensis -NOCD "${{ github.workspace }}/${{ env.BUILD_DIR }}/program_info/win_install.nsi"
```
Comment (DioEgizio) [inline .github/workflows/build.yml]: space that wasn't there before

**Comment by DioEgizio** [comment]
also

>  Build AppImage using Qt 6

You might wanna tick that in the description 🙂

File: launcher/net/PasteUpload.cpp (hunk 2 above)
Comment (Scrumplex) [reply launcher/net/PasteUpload.cpp]: it is possible that this is duplicate yes.

File: launcher/ui/MainWindow.cpp (hunk 3 above)
Comment (Scrumplex) [reply launcher/ui/MainWindow.cpp:238]: it was ununsed so i removed it ¯\\\_(ツ)\_/¯

File: .github/workflows/build.yml (hunk 5 above)
Comment (Scrumplex) [reply .github/workflows/build.yml]: NOOOOOO

Comment (DioEgizio) [reply .github/workflows/build.yml]: YEEEES :pofat:

**Comment by Scrumplex** [comment]
> those (c) diffs eye . eye

copyright headers should all be in a single commit, so if you manage to exclude that one while reviewing it should be a bit more readable :D

**Comment by Scrumplex** [comment]
Making this a draft again.

File: launcher/ui/MainWindow.cpp (hunk 3 above)
Comment (flowln) [reply launcher/ui/MainWindow.cpp:238]: erm i guess we can add it later :|

File: launcher/ui/dialogs/SkinUploadDialog.cpp (hunk 4 above)
Comment (flowln) [reply launcher/ui/dialogs/SkinUploadDialog.cpp]:
```suggestion
    if(urlPrefixMatcher.match(input).hasMatch())
```
yes

**Review by flowln — APPROVED** [review]
lgtm! nice job!! /gen

it's working very fine in Qt5 too :+1:

**Review by txtsd — APPROVED** [review]
Runs well for me!

Can't believe I thought I could tackle the qt5->qt6 change solo at one point :sweat_smile:

**Comment by Scrumplex** [comment]
> Can't believe I thought I could tackle the qt5->qt6 change solo at one point 😅

I mean I managed to tackle it solo mostly :D

It was mostly straightforward. There were just many places that needed changes

**Comment by txtsd** [comment]
You're definitely more capable than I am, at least as far CMake and C++ are concerned :grinning:

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -16,20 +16,31 @@ jobs:
...
           - os: ubuntu-20.04
             appimage: true
+            qt_ver: 6
+            qt_host: linux
+            qt_version: '6.3.1'
+            qt_modules: 'qt5compat qtimageformats'
 
           - os: windows-2022
-            name: "Windows-i686"
+            name: "Windows-i686-Legacy"
```
Comment (DioEgizio) [inline .github/workflows/build.yml]:
```suggestion
            name: "Windows-Legacy"
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -16,20 +16,31 @@ jobs:
...
             msystem: mingw32
+            qt_ver: 5
 
           - os: windows-2022
-            name: "Windows-x86_64"
-            msystem: mingw64
+            name: "Windows-i686"
```
Comment (DioEgizio) [inline .github/workflows/build.yml]:
```suggestion
            name: "Windows"
```

**Comment by txtsd** [comment]
Dang, builds are still failing.

**Comment by TayouVR** [comment]
in the in-app about section and in the repo the license info would have to be updated to state Qt 6, wouldn't it?
image

**Comment by DioEgizio** [comment]
Yeah COPYING should be updated in general, that portion is just wrong because qt5 is lgpl3 anyways

**Review by d-513 — APPROVED** [review]

**Comment by Scrumplex** [comment]
Can someone on macOS test builds from here?

**Comment by Scrumplex** [comment]
Looks like it's working fine

**Comment by DioEgizio** [comment]
all conversation resolved, 3 approvals
Merge when?

**Pull request merged**